core/src/main/java/org/elasticsearch/action/ingest/IngestActionFilter.java

```diff
@@ -0,0 +1,227 @@
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.bulk.BulkAction;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.index.IndexAction;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.support.ActionFilter;
+import org.elasticsearch.action.support.ActionFilterChain;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.ingest.PipelineExecutionService;
+import org.elasticsearch.node.service.NodeService;
+import org.elasticsearch.tasks.Task;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+public final class IngestActionFilter extends AbstractComponent implements ActionFilter {
+
+    private final PipelineExecutionService executionService;
+
+    @Inject
+    public IngestActionFilter(Settings settings, NodeService nodeService) {
+        super(settings);
+        this.executionService = nodeService.getIngestService().getPipelineExecutionService();
+    }
+
+    @Override
+    public void apply(Task task, String action, ActionRequest<?> request, ActionListener<?> listener, ActionFilterChain chain) {
+        if (IndexAction.NAME.equals(action)) {
+            assert request instanceof IndexRequest;
+            IndexRequest indexRequest = (IndexRequest) request;
+            if (Strings.hasText(indexRequest.pipeline())) {
+                processIndexRequest(task, action, listener, chain, (IndexRequest) request);
+                return;
+            }
+        }
+        if (BulkAction.NAME.equals(action)) {
+            assert request instanceof BulkRequest;
+            BulkRequest bulkRequest = (BulkRequest) request;
+            boolean isIngestRequest = false;
+            for (ActionRequest actionRequest : bulkRequest.requests()) {
+                if (actionRequest instanceof IndexRequest) {
+                    IndexRequest indexRequest = (IndexRequest) actionRequest;
+                    if (Strings.hasText(indexRequest.pipeline())) {
+                        isIngestRequest = true;
+                        break;
+                    }
+                }
+            }
+            if (isIngestRequest) {
+                @SuppressWarnings("unchecked")
+                ActionListener<BulkResponse> actionListener = (ActionListener<BulkResponse>) listener;
+                processBulkIndexRequest(task, bulkRequest, action, chain, actionListener);
+                return;
+            }
+        }
+
+        chain.proceed(task, action, request, listener);
```

Comment:
this should also be in an else clause and then we can drop all the return statements?